drivers/led/lp503x.c

```diff
@@ -0,0 +1,286 @@
+			   const struct led_info **info)
+{
+	const struct lp503x_config *config = DEV_CFG(dev);
+	const struct led_info *led_info = lp503x_led_to_info(config, led);
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	*info = led_info;
+
+	return 0;
+}
+
+static int lp503x_set_brightness(struct device *dev,
+				 uint32_t led, uint8_t value)
+{
+	const struct lp503x_config *config = DEV_CFG(dev);
+	struct lp503x_data *data = DEV_DATA(dev);
+	const struct led_info *led_info = lp503x_led_to_info(config, led);
+	uint8_t buf[2];
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	buf[0] = LP503X_LED_BRIGHTNESS_BASE + led_info->index;
+	buf[1] = value;
+
+	return i2c_write(data->i2c, buf, sizeof(buf), config->i2c_addr);
+}
+
+static int lp503x_on(struct device *dev, uint32_t led)
+{
+	return lp503x_set_brightness(dev, led, 255);
+}
+
+static int lp503x_off(struct device *dev, uint32_t led)
+{
+	return lp503x_set_brightness(dev, led, 0);
+}
+
+static int lp503x_set_color(struct device *dev, uint32_t led, uint8_t *color)
+{
+	const struct lp503x_config *config = DEV_CFG(dev);
+	struct lp503x_data *data = DEV_DATA(dev);
+	const struct led_info *led_info = lp503x_led_to_info(config, led);
+	uint8_t buf[4];
+
+	if (!led_info) {
+		return -EINVAL;
+	}
+
+	buf[0] = LP503X_OUT_COLOR_BASE + 3 * led_info->index;
+	buf[1] = color[0];
+	buf[2] = color[1];
+	buf[3] = color[2];
+
+	return i2c_write(data->i2c, buf, sizeof(buf), config->i2c_addr);
+}
+
+static int lp503x_write_channels(struct device *dev, uint32_t start_channel,
+				 uint32_t num_channels, uint8_t *buf)
+{
+	const struct lp503x_config *config = DEV_CFG(dev);
+	struct lp503x_data *data = DEV_DATA(dev);
+
+	if (start_channel >= LP503X_NUM_CHANNELS ||
+	    start_channel + num_channels > LP503X_NUM_CHANNELS) {
+		return -EINVAL;
+	}
+
+	/*
+	 * Unfortunately this controller don't support commands split into
+	 * two I2C messages.
+	 */
+	data->chan_buf[0] = LP503X_CHANNEL_BASE + start_channel;
+	memcpy(data->chan_buf + 1, buf, num_channels);
+
+	return i2c_write(data->i2c, data->chan_buf,
+			 num_channels + 1, config->i2c_addr);
+}
+
+static int lp503x_init(struct device *dev)
+{
+	const struct lp503x_config *config = DEV_CFG(dev);
+	struct lp503x_data *data = DEV_DATA(dev);
+	uint8_t buf[3];
+	int err;
+
+	data->i2c = device_get_binding(config->i2c_bus_label);
+	if (data->i2c == NULL) {
+		LOG_ERR("%s: device %s not found",
+			dev->name, config->i2c_bus_label);
+		return -ENODEV;
+	}
+	if (config->num_leds > LP503X_MAX_LEDS) {
+		LOG_ERR("%s: invalid number of LEDs %d (max %d)",
+			dev->name, config->num_leds, LP503X_MAX_LEDS);
+		return -EINVAL;
+	}
+
+	/*
+	 * Since the status of the LP503x controller is unknown when entering
+	 * this function, and since there is no way to reset it, then the whole
+	 * configuration must be applied.
+	 */
+
+	/* Disable bank control for all LEDs. */
+	buf[0] = LP503X_LED_CONFIG0;
+	buf[1] = 0;
+	buf[2] = 0;
+	err = i2c_write(data->i2c, buf, 2, config->i2c_addr);
```

Comment:
num_bytes is 2, not 3 or `sizeof(buf)`?

Same question for the rest of the function.

Reply:
Yes it should be 3 here, it is a bug. Both the LP503X_LED_CONFIG0 and LP503X_LED_CONFIG1 registers are written. Thanks for spotting that. Both the next calls (using 2) are corrects:  a single register is written.